File: ios/ClingSyncTests/RecentTargetsTest.swift
import Foundation
import Testing

@testable import ClingSync

struct RecentTargetsTest {
    private func freshDefaults() -> UserDefaults {
        let defaults = UserDefaults(suiteName: "recent-targets-test")!
        defaults.removePersistentDomain(forName: "recent-targets-test")
        return defaults
    }

    @Test func recordsMostRecentFirstDedupedAndCapped() {
        let defaults = freshDefaults()
        for index in 0..<12 {
            RecentTargets.record("dir\(index)", defaults)
        }
        let list = RecentTargets.load(defaults)
        #expect(list.count == RecentTargets.maxCount)
        #expect(list.first == "dir11")

        RecentTargets.record("dir5", defaults)
        let moved = RecentTargets.load(defaults)
        #expect(moved.first == "dir5")
        #expect(moved.filter { $0 == "dir5" }.count == 1)
        #expect(moved.count == RecentTargets.maxCount)
    }

    @Test func recordTrimsWhitespace() {
        let defaults = freshDefaults()
        RecentTargets.record("  photos/2024  ", defaults)
        #expect(RecentTargets.load(defaults) == ["photos/2024"])
    }

    @Test func recordNormalizesSurroundingSlashesSoEquivalentPathsCollapse() {
        let defaults = freshDefaults()
        RecentTargets.record("/photos/2024/", defaults)
        RecentTargets.record("photos/2024", defaults)
        #expect(RecentTargets.load(defaults) == ["photos/2024"])
    }

    @Test func optionsPutRecentFirstThenSettingsPrefixWithLastUsedDefault() {
        let options = ShareTargetOptions(settingsPrefix: "backup", recent: ["photos", "docs"])
        #expect(options.options == ["photos", "docs", "backup"])
        #expect(options.defaultTarget == "photos")
    }

    @Test func optionsWithoutRecentDefaultToSettingsPrefix() {
        let options = ShareTargetOptions(settingsPrefix: "backup", recent: [])
        #expect(options.options == ["backup"])
        #expect(options.defaultTarget == "backup")
    }

    @Test func optionsDoNotDuplicateSettingsPrefixAlreadyInRecent() {
        let options = ShareTargetOptions(settingsPrefix: "photos", recent: ["photos", "docs"])
        #expect(options.options == ["photos", "docs"])
        #expect(options.defaultTarget == "photos")
    }
}

File: android/app/src/test/java/com/clingsync/android/RecentTargetsTest.kt
package com.clingsync.android

import android.content.Context
import androidx.test.core.app.ApplicationProvider
import androidx.test.ext.junit.runners.AndroidJUnit4
import org.junit.Assert.assertEquals
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.annotation.Config

@RunWith(AndroidJUnit4::class)
@Config(sdk = [28])
class RecentTargetsTest {
    private lateinit var context: Context

    @Before
    fun setup() {
        context = ApplicationProvider.getApplicationContext()
        context.getSharedPreferences("recent_share_targets", Context.MODE_PRIVATE).edit().clear().commit()
    }

    @Test
    fun recordsMostRecentFirstDedupedAndCapped() {
        repeat(12) { RecentTargets.record(context, "dir$it") }

        val list = RecentTargets.load(context)
        assertEquals(RecentTargets.MAX, list.size)
        assertEquals("dir11", list.first())

        RecentTargets.record(context, "dir5")
        val moved = RecentTargets.load(context)
        assertEquals("dir5", moved.first())
        assertEquals(1, moved.count { it == "dir5" })
        assertEquals(RecentTargets.MAX, moved.size)
    }

    @Test
    fun recordTrimsWhitespace() {
        RecentTargets.record(context, "  photos/2024  ")
        assertEquals(listOf("photos/2024"), RecentTargets.load(context))
    }

    @Test
    fun recordNormalizesSurroundingSlashesSoEquivalentPathsCollapse() {
        RecentTargets.record(context, "/photos/2024/")
        RecentTargets.record(context, "photos/2024")
        assertEquals(listOf("photos/2024"), RecentTargets.load(context))
    }

    @Test
    fun optionsPutRecentFirstThenSettingsPrefixWithLastUsedDefault() {
        val options = ShareTargetOptions.from(settingsPrefix = "backup", recent = listOf("photos", "docs"))
        assertEquals(listOf("photos", "docs", "backup"), options.options)
        assertEquals("photos", options.default)
    }

    @Test
    fun optionsWithoutRecentDefaultToSettingsPrefix() {
        val options = ShareTargetOptions.from(settingsPrefix = "backup", recent = emptyList())
        assertEquals(listOf("backup"), options.options)
        assertEquals("backup", options.default)
    }

    @Test
    fun optionsDoNotDuplicateSettingsPrefixAlreadyInRecent() {
        val options = ShareTargetOptions.from(settingsPrefix = "photos", recent = listOf("photos", "docs"))
        assertEquals(listOf("photos", "docs"), options.options)
        assertEquals("photos", options.default)
    }
}
